File: src/org/jgll/util/hashing/CuckooHashMap.java
package org.jgll.util.hashing;

import java.io.Serializable;

/**
 * 
 * A hash set based on Cuckoo hashing.
 * 
 * @author Ali Afroozeh
 *
 */
public class CuckooHashMap<K, V> implements Serializable {
	
	private static final long serialVersionUID = 1L;
	
	private CuckooHashSet<MapEntry<K, V>> set;

	public CuckooHashMap(ExternalHasher<K> decomposer) {
		set = new CuckooHashSet<>(new MapEntryExternalHasher(decomposer));
	}
	
	public CuckooHashMap(int initialCapacity, ExternalHasher<K> decomposer) {
		set = new CuckooHashSet<>(initialCapacity, new MapEntryExternalHasher(decomposer));
	}
	
	public V get(K key) {
		MapEntry<K, V> entry = set.get(new MapEntry<K, V>(key, null));
		if(entry != null) {
			return entry.getValue();
		}
		
		return null;
	}
	
	public V put(K key, V value) {
		MapEntry<K, V> add = set.add(new MapEntry<K, V>(key, value));
		if(add == null) {
			return null;
		}
		
		return add.v;
	}
	
	public int size() {
		return set.size();
	}
	
	public void clear() {
		set.clear();
	}
	
	public static class MapEntry<K, V> {
		
		private K k;
		private V v;
		
		public MapEntry(K k, V v) {
			this.k = k;
			this.v = v;
		}

		public K getKey() {
			return k;
		}

		public V getValue() {
			return v;
		}

		public V setValue(V value) {
			this.v = value;
			return v;
		}
		
		@Override
		public boolean equals(Object obj) {
			if(this == obj) {
				return true;
			}
			
			if(! (obj instanceof MapEntry)) {
				return false;
			}
			
			@SuppressWarnings("unchecked")
			MapEntry<K, V> other = (MapEntry<K, V>) obj;
			
			return k.equals(other.k);
		}
		
		@Override
		public String toString() {
			return "(" + k.toString() + ", " + (v == null ? "" : v.toString()) + ")";
		}
	}
	
	public class MapEntryExternalHasher implements ExternalHasher<MapEntry<K, V>> {

		private static final long serialVersionUID = 1L;
		
		private ExternalHasher<K> hasher;

		public MapEntryExternalHasher(ExternalHasher<K> hasher) {
			this.hasher = hasher;
		}
		
		@Override
		public int hash(MapEntry<K, V> t, HashFunction f) {
			return hasher.hash(t.k, f);
		}
	}
}

File: src-scala/org/jgll/util/hashing/CuckooHashMap.scala
package org.jgll.util.hashing

import java.io.Serializable
import org.jgll.util.hashing.hashfunction.HashFunction
import CuckooHashMap._

object CuckooHashMap {

  class MapEntry[K: Manifest, V: Manifest](var k: K, var v: V) {

    def getKey(): K = k

    def getValue(): V = v

    def setValue(value: V): V = {
      this.v = value
      v
    }

    override def equals(obj: Any): Boolean = {
      if (!(obj.isInstanceOf[MapEntry[K,V]])) {
        return false
      }
      val other = obj.asInstanceOf[MapEntry[K, V]]
      k == other.k
    }

    override def toString(): String = {
      "(" + k.toString + ", " + (if (v == null) "" else v.toString) +
        ")"
    }
  }
}

@SerialVersionUID(1L)
class CuckooHashMap[K, V >: Null](decomposer: ExternalHasher[K], initialCapacity: Int = CuckooHashSet.DEFAULT_INITIAL_CAPACITY)
                                 (implicit val kmanifest: Manifest[K], implicit val vmanifest: Manifest[V])
  extends Serializable {

  private var set: CuckooHashSet[MapEntry[K, V]] = new CuckooHashSet(initialCapacity, new MapEntryExternalHasher(decomposer))

  def get(key: K): V = {
    val entry = set.get(new MapEntry[K, V](key, null))
    if (entry != null) {
      return entry.getValue
    }
    null
  }

  def put(key: K, value: V): V = {
    val add = set.add(new MapEntry[K, V](key, value))
    if (add == null) {
      return null
    }
    add.v
  }

  def size(): Int = set.size

  def clear() {
    set.clear()
  }

  @SerialVersionUID(1L)
  class MapEntryExternalHasher(private var hasher: ExternalHasher[K]) extends ExternalHasher[MapEntry[K, V]] {

    override def hash(t: MapEntry[K, V], f: HashFunction): Int = hasher.hash(t.k, f)

    override def equals(e1: MapEntry[K, V], e2: MapEntry[K, V]): Boolean = hasher.equals(e1.k, e2.k)
  }
}
